lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,292 @@
+		logger.Debug("CDI 'chmod' hook is ignored")
+		return nil
+	default:
+		return fmt.Errorf("Unsupported CDI hook function %q", hook.Args[2])
+	}
+
+	return nil
+}
+
+// GenerateFromCDI does several things:
+//
+//  1. It generates a CDI specification from a CDI ID and an instance.
+//     According the the specified 'vendor', 'class' and 'name' (this assembled triplet is called a fully-qualified CDI ID. We'll just call it ID in the context of this package), the CDI specification is generated.
+//     The CDI specification is a JSON-like format. It is divided into two parts: the 'specific device' configuration and the 'general device' configuration.
+//     - The 'specific device' configuration: this is a list of 'container edits' that can be added to the container runtime.
+//     According to the CDI ID (vendor, class, name), we only select the 'container edits' that matches the CDI ID.
+//     The 'container edits' are a list of device nodes, hooks and mounts that must be added to the container runtime.
+//     - The 'general device' configuration: this is a single 'container edits' entry runtime that must be passed to the container runtime in ant case. Which unix char devices need to be passed
+//     (e.g, special GPU memory controller device, etc.)? Which user space libraries need to be mounted (e.g, CUDA libraries for NVIDIA, etc.)?
+//     Which hooks need to be executed (e.g, symlinks to create, folder entries to add to ldcache, etc.))?
+//     In our case, these edits will be interpreted either as disk or unix char mounts passed to the container.
+//     The hooks will be centralized in a single resource file that will be read and executed as a LXC `lxc.hook.mount` hook,
+//     through LXD's `callhook` command.
+//  2. We first process the 'specific device' configuration: we convert this information into a map of devices
+//     (keyed by their path given in the spec, it mapped to a map of device properties). We also collect the specific mounts (but we do not process them yet) and hooks.
+//  3. We then process the 'general device' configuration in the same fashion.
+//  4. Now we process all the mounts we collected from the spec in order to turn them into disk devices.
+//     This operations generate a side effect: it generates a list of indirect symlinks (see `specMountToNativeDev`)
+//  5. Merge all the hooks (direct + indirect) into a single list of hooks.
+func GenerateFromCDI(inst instance.Instance, cdiID ID, l logger.Logger) (*ConfigDevices, *Hooks, error) {
+	// 1. Generate the CDI specification
+	spec, err := generateSpec(cdiID, inst, l)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate CDI spec: %w", err)
+	}
+
+	var targetName string
+	cdiName := cdiID.Name
+	if len(cdiName) == 1 {
+		targetName = cdiName[0]
+	} else if len(cdiName) == 2 && cdiID.Class == MIG {
+		targetName = fmt.Sprintf("%s:%s", cdiName[0], cdiName[1])
+	} else {
+		return nil, nil, fmt.Errorf("Invalid CDI locator %v", cdiName)
```

Comment:
if cdiName is a string then should use %q not %v

Reply:
fixed